File: e2e/generate-report.py
"""Generate audit report from Playwright JSON results."""
import json, sys, os
from datetime import datetime

sys.stdout.reconfigure(encoding='utf-8')

RESULTS = os.path.join(os.path.dirname(__file__), 'results.json')
OUTPUT = os.path.join(os.path.dirname(__file__), '..', 'DEMO-AUDIT-REPORT.md')

with open(RESULTS, encoding='utf-8') as f:
    data = json.load(f)

# Extract all tests
tests = []

def walk_suites(suite, section=''):
    sec = suite.get('title') or section
    for spec in suite.get('specs', []):
        for t in spec.get('tests', []):
            result = t['results'][0] if t.get('results') else {}
            annotations = result.get('annotations', [])
            issues = []
            for ann in annotations:
                if ann.get('type') == 'issues' and ann.get('description'):
                    issues = json.loads(ann['description'])
            tests.append({
                'section': sec,
                'name': spec['title'],
                'status': result.get('status', 'unknown'),
                'duration': result.get('duration', 0),
                'issues': issues,
                'errors': result.get('errors', []),
            })
    for child in suite.get('suites', []):
        walk_suites(child, sec)

for s in data['suites']:
    walk_suites(s)

# Group by section
from collections import OrderedDict
sections = OrderedDict()
for t in tests:
    sections.setdefault(t['section'], []).append(t)

# Build report
L = []
L.append('# Bloom House — Automated Demo Audit Report')
L.append(f'**Generated:** {datetime.now().strftime("%Y-%m-%d %H:%M")}')
L.append(f'**Target:** https://bloom-house-iota.vercel.app/demo/*')
L.append(f'**Method:** Playwright Chromium — full JS rendering, network monitoring, console capture')
L.append(f'**Total duration:** {data["stats"]["duration"]/1000:.0f}s')
L.append('')

total = len(tests)
passed = sum(1 for t in tests if t['status'] == 'passed')
failed = total - passed
with_issues = sum(1 for t in tests if len(t['issues']) > 0)
all_issues = [i for t in tests for i in t['issues']]

L.append('## Summary')
L.append('| Metric | Count |')
L.append('|--------|-------|')
L.append(f'| Total pages tested | {total} |')
L.append(f'| Passed (no critical issues) | {passed} |')
L.append(f'| Failed (critical issues) | {failed} |')
L.append(f'| Pages with warnings | {with_issues} |')
L.append(f'| Total issues found | {len(all_issues)} |')
L.append('')

L.append('## What Was Checked')
L.append('Each page was loaded in a real Chromium browser with full JavaScript execution:')
L.append('')
L.append('| Check | What it catches |')
L.append('|-------|----------------|')
L.append('| **Console errors** | Runtime JS crashes, unhandled promise rejections, failed imports |')
L.append('| **Bad text patterns** | Visible "undefined", "NaN", "null", "[object Object]", error messages, "Lorem ipsum" |')
L.append('| **Failed network requests** | Broken API calls (4xx/5xx), missing endpoints |')
L.append('| **Broken images** | Images that failed to load (naturalWidth === 0) |')
L.append('| **Empty pages** | Pages with < 20 characters of visible text (blank renders) |')
L.append('| **Screenshots** | Every page screenshotted for visual review |')
L.append('')

# Issue type counts
type_counts = {}
for i in all_issues:
    type_counts[i['type']] = type_counts.get(i['type'], 0) + 1

if type_counts:
    L.append('## Issue Breakdown by Type')
    L.append('| Type | Count | Severity |')
    L.append('|------|-------|----------|')
    severity = {
        'console-error': 'Medium',
        'bad-text': 'High',
        'failed-request': 'Medium-High',
        'broken-image': 'Medium',
        'empty-page': 'Critical',
    }
    for typ, cnt in sorted(type_counts.items(), key=lambda x: -x[1]):
        L.append(f'| {typ} | {cnt} | {severity.get(typ, "Unknown")} |')
    L.append('')

# Per-section detail
L.append('---')
L.append('')
L.append('## Detailed Results by Section')
L.append('')

for section, stests in sections.items():
    sp = sum(1 for t in stests if t['status'] == 'passed')
    sf = len(stests) - sp
    si = sum(1 for t in stests if t['issues'])

    status_str = f'{sp} passed'
    if sf > 0: status_str += f', {sf} failed'
    if si > 0: status_str += f', {si} with warnings'

    L.append(f'### {section} ({len(stests)} pages — {status_str})')
    L.append('')

    L.append('| Page | Status | Time | Issues |')
    L.append('|------|--------|------|--------|')
    for t in stests:
        icon = '✅' if t['status'] == 'passed' else '❌'
        dur = f'{t["duration"]/1000:.1f}s'
        ic = f'⚠️ {len(t["issues"])}' if t['issues'] else '—'
        L.append(f'| `{t["name"]}` | {icon} {t["status"]} | {dur} | {ic} |')
    L.append('')

    # Issue details for this section
    issues_in_section = [(t['name'], t['issues']) for t in stests if t['issues']]
    if issues_in_section:
        L.append(f'#### {section} — Issue Details')
        L.append('')
        for name, issues in issues_in_section:
            L.append(f'**`{name}`:**')
            for issue in issues:
                icon = {
                    'console-error': '🔴',
                    'bad-text': '🟡',
                    'failed-request': '🔴',
                    'broken-image': '🟠',
                    'empty-page': '⛔',
                }[issue['type']]
                L.append(f'- {icon} **{issue["type"]}**: {issue["detail"]}')
            L.append('')

# Deduplicate recurring issues
L.append('---')
L.append('')
L.append('## Recurring Issues (Deduplicated)')
L.append('')
L.append('These issues appear across multiple pages and likely share a root cause:')
L.append('')

from collections import Counter
issue_details = Counter()
issue_pages = {}
for t in tests:
    for i in t['issues']:
        key = (i['type'], i['detail'][:120])
        issue_details[key] += 1
        issue_pages.setdefault(key, []).append(t['name'])

recurring = [(k, v) for k, v in issue_details.items() if v >= 2]
recurring.sort(key=lambda x: -x[1])

if recurring:
    L.append('| # | Type | Detail | Pages Affected | Count |')
    L.append('|---|------|--------|----------------|-------|')
    for idx, ((typ, detail), count) in enumerate(recurring, 1):
        pages = issue_pages[(typ, detail)]
        page_str = ', '.join(pages[:5])
        if len(pages) > 5:
            page_str += f' +{len(pages)-5} more'
        L.append(f'| {idx} | `{typ}` | {detail} | {page_str} | {count} |')
    L.append('')
else:
    L.append('No recurring issues found.')
    L.append('')

# Clean pages
L.append('---')
L.append('')
L.append('## Clean Pages (No Issues Detected)')
L.append('')
clean = [t for t in tests if t['status'] == 'passed' and not t['issues']]
if clean:
    for t in clean:
        L.append(f'- ✅ {t["section"]} / `{t["name"]}`')
else:
    L.append('No completely clean pages — every page had at least one warning.')
L.append('')

# Manual review notes
L.append('---')
L.append('')
L.append('## Still Needs Manual Review')
L.append('')
L.append('Automated testing catches structural issues. These still need human eyes:')
L.append('')
L.append('- [ ] Charts/graphs render with real data (not empty containers)')
L.append('- [ ] Drag-and-drop interactions work (seating, timeline reordering)')
L.append('- [ ] Form submissions save and persist correctly')
L.append('- [ ] Sage AI chat responds coherently with venue voice')
L.append('- [ ] Cross-page data consistency (portal config ↔ couple portal display)')
L.append('- [ ] Mobile/tablet responsiveness across breakpoints')
L.append('- [ ] Correct venue branding per demo venue (not just Hawthorne Manor)')
L.append('- [ ] Print/PDF export functionality')
L.append('- [ ] Email notifications trigger correctly')
L.append('- [ ] Multi-venue scope switching works (Hawthorne → Crestwood → Glass House → Rose Hill)')
L.append('')
L.append('---')
L.append(f'*Screenshots saved to `test-results/` — one per page for visual walkthrough.*')

report = '\n'.join(L)
with open(OUTPUT, 'w', encoding='utf-8') as f:
    f.write(report)

print(f'Report written to {OUTPUT}')
print(f'{total} pages, {passed} passed, {failed} failed, {with_issues} with warnings, {len(all_issues)} total issues')

File: e2e/generate-report.ts
/**
 * Reads Playwright JSON results + screenshots and generates a single
 * markdown audit report ready to paste into Claude.
 *
 * Run: npx tsx e2e/generate-report.ts
 */
import * as fs from 'fs'
import * as path from 'path'

interface Issue {
  type: 'console-error' | 'bad-text' | 'failed-request' | 'broken-image' | 'empty-page'
  detail: string
}

interface TestResult {
  title: string
  fullTitle: string
  status: 'passed' | 'failed' | 'timedOut' | 'skipped'
  duration: number
  annotations: { type: string; description?: string }[]
  errors: { message: string }[]
}

interface Suite {
  title: string
  suites?: Suite[]
  specs?: { title: string; tests: TestResult[] }[]
}

interface PlaywrightReport {
  suites: Suite[]
  stats: {
    expected: number
    unexpected: number
    flaky: number
    skipped: number
    duration: number
  }
}

function extractTests(suite: Suite, section = ''): { section: string; name: string; result: TestResult }[] {
  const out: { section: string; name: string; result: TestResult }[] = []
  const currentSection = suite.title || section

  for (const spec of suite.specs || []) {
    for (const t of spec.tests) {
      out.push({ section: currentSection, name: spec.title, result: t })
    }
  }
  for (const child of suite.suites || []) {
    out.push(...extractTests(child, currentSection))
  }
  return out
}

function main() {
  const resultsPath = path.join(__dirname, 'results.json')
  if (!fs.existsSync(resultsPath)) {
    console.error('No results.json found. Run tests first: npx playwright test')
    process.exit(1)
  }

  const report: PlaywrightReport = JSON.parse(fs.readFileSync(resultsPath, 'utf-8'))
  const allTests = report.suites.flatMap(s => extractTests(s))

  // Group by section
  const sections = new Map<string, typeof allTests>()
  for (const t of allTests) {
    const arr = sections.get(t.section) || []
    arr.push(t)
    sections.set(t.section, arr)
  }

  const lines: string[] = []
  lines.push('# Bloom House — Automated Demo Audit Report')
  lines.push(`**Generated:** ${new Date().toISOString().split('T')[0]}`)
  lines.push(`**Target:** https://bloom-house-iota.vercel.app/demo/*`)
  lines.push(`**Method:** Playwright Chromium — full JS rendering, network monitoring, console capture`)
  lines.push('')

  // Summary
  const total = allTests.length
  const passed = allTests.filter(t => t.result.status === 'passed').length
  const failed = allTests.filter(t => t.result.status !== 'passed').length
  const withIssues = allTests.filter(t => {
    const ann = t.result.annotations.find(a => a.type === 'issues')
    return ann && ann.description && JSON.parse(ann.description).length > 0
  }).length

  lines.push('## Summary')
  lines.push(`| Metric | Count |`)
  lines.push(`|--------|-------|`)
  lines.push(`| Total pages tested | ${total} |`)
  lines.push(`| Passed (no critical issues) | ${passed} |`)
  lines.push(`| Failed (critical issues) | ${failed} |`)
  lines.push(`| Pages with warnings | ${withIssues} |`)
  lines.push('')

  // What was checked
  lines.push('## What Was Checked')
  lines.push('Each page was loaded in a real Chromium browser with full JavaScript execution. The following checks ran automatically:')
  lines.push('')
  lines.push('| Check | What it catches |')
  lines.push('|-------|----------------|')
  lines.push('| **Console errors** | Runtime JS crashes, unhandled promise rejections, failed imports |')
  lines.push('| **Bad text patterns** | Visible "undefined", "NaN", "null", "[object Object]", error messages, "Lorem ipsum" |')
  lines.push('| **Failed network requests** | Broken API calls (4xx/5xx), missing endpoints |')
  lines.push('| **Broken images** | Images that failed to load (naturalWidth === 0) |')
  lines.push('| **Empty pages** | Pages with < 20 characters of visible text (blank renders) |')
  lines.push('| **Screenshots** | Every page screenshotted for visual review |')
  lines.push('')

  // Issue type counts
  const allIssues: { section: string; page: string; issues: Issue[] }[] = []
  for (const t of allTests) {
    const ann = t.result.annotations.find(a => a.type === 'issues')
    if (ann?.description) {
      const issues: Issue[] = JSON.parse(ann.description)
      if (issues.length > 0) {
        allIssues.push({ section: t.section, page: t.name, issues })
      }
    }
  }

  const issueCounts: Record<string, number> = {}
  for (const { issues } of allIssues) {
    for (const i of issues) {
      issueCounts[i.type] = (issueCounts[i.type] || 0) + 1
    }
  }

  if (Object.keys(issueCounts).length > 0) {
    lines.push('## Issue Breakdown by Type')
    lines.push('| Type | Count |')
    lines.push('|------|-------|')
    for (const [type, count] of Object.entries(issueCounts).sort((a, b) => b[1] - a[1])) {
      lines.push(`| ${type} | ${count} |`)
    }
    lines.push('')
  }

  // Per-section detail
  lines.push('---')
  lines.push('')
  lines.push('## Detailed Results by Section')
  lines.push('')

  for (const [section, tests] of sections) {
    const sectionPassed = tests.filter(t => t.result.status === 'passed').length
    const sectionFailed = tests.length - sectionPassed
    const sectionIssues = allIssues.filter(i => i.section === section)

    lines.push(`### ${section} (${tests.length} pages — ${sectionPassed} passed, ${sectionFailed} failed)`)
    lines.push('')

    // Table of all pages
    lines.push('| Page | Status | Duration | Issues |')
    lines.push('|------|--------|----------|--------|')
    for (const t of tests) {
      const status = t.result.status === 'passed' ? '✅' : '❌'
      const duration = `${(t.result.duration / 1000).toFixed(1)}s`
      const ann = t.result.annotations.find(a => a.type === 'issues')
      const issues: Issue[] = ann?.description ? JSON.parse(ann.description) : []
      const issueCount = issues.length > 0 ? `⚠️ ${issues.length}` : '—'
      lines.push(`| ${t.name} | ${status} ${t.result.status} | ${duration} | ${issueCount} |`)
    }
    lines.push('')

    // Detail for pages with issues
    if (sectionIssues.length > 0) {
      lines.push(`**Issues found in ${section}:**`)
      lines.push('')
      for (const { page, issues } of sectionIssues) {
        lines.push(`**${page}:**`)
        for (const issue of issues) {
          const icon = {
            'console-error': '🔴',
            'bad-text': '🟡',
            'failed-request': '🔴',
            'broken-image': '🟠',
            'empty-page': '🔴',
          }[issue.type] || '⚪'
          lines.push(`- ${icon} \`${issue.type}\`: ${issue.detail}`)
        }
        lines.push('')
      }
    }

    // Detail for failed tests
    const failedTests = tests.filter(t => t.result.status !== 'passed')
    if (failedTests.length > 0) {
      for (const t of failedTests) {
        if (t.result.errors.length > 0) {
          lines.push(`**${t.name} — error:**`)
          lines.push('```')
          lines.push(t.result.errors[0].message.slice(0, 500))
          lines.push('```')
          lines.push('')
        }
      }
    }
  }

  // Clean pages list
  lines.push('---')
  lines.push('')
  lines.push('## Clean Pages (No Issues Detected)')
  lines.push('')
  const cleanPages = allTests.filter(t => {
    const ann = t.result.annotations.find(a => a.type === 'issues')
    const issues: Issue[] = ann?.description ? JSON.parse(ann.description) : []
    return t.result.status === 'passed' && issues.length === 0
  })
  if (cleanPages.length > 0) {
    for (const t of cleanPages) {
      lines.push(`- ✅ ${t.section} / ${t.name}`)
    }
  } else {
    lines.push('No completely clean pages — every page had at least one warning.')
  }

  lines.push('')
  lines.push('---')
  lines.push('')
  lines.push('## Notes for Manual Review')
  lines.push('')
  lines.push('This automated audit catches **structural** issues only. The following still need manual verification:')
  lines.push('- Charts/graphs render with actual data (not empty containers)')
  lines.push('- Drag-and-drop interactions work (seating, timeline)')
  lines.push('- Form submissions save correctly')
  lines.push('- Sage AI chat responds coherently')
  lines.push('- Cross-page data consistency (portal config → couple portal display)')
  lines.push('- Mobile responsiveness')
  lines.push('- Correct venue branding per demo venue')
  lines.push('- Print/PDF export functionality')
  lines.push('')
  lines.push(`*Screenshots saved to: test-results/ directory*`)

  const reportText = lines.join('\n')
  const outPath = path.join(__dirname, '..', 'DEMO-AUDIT-REPORT.md')
  fs.writeFileSync(outPath, reportText, 'utf-8')
  console.log(`Report written to ${outPath}`)
  console.log(`${total} pages tested, ${passed} passed, ${failed} failed, ${withIssues} with warnings`)
}

main()
